src/lib.rs

```diff
@@ -1021,59 +1023,95 @@ fn get_docker_for_tcp(tcp_host_str: String) -> Docker {
-    Docker {
+    Ok(Docker {
         transport: Transport::Tcp {
             client: Client::builder().build(http),
             host: tcp_host_str,
         },
+    })
+}
+
+impl Default for Docker {
+    /// Returns a Docker instance connected to a unix socket at [DEFAULT_SOCKET_PATH](DEFAULT_SOCKET_PATH)
+    fn default() -> Self {
+        Docker::unix(DEFAULT_SOCKET_PATH)
     }
 }
 
 // https://docs.docker.com/reference/api/docker_remote_api_v1.17/
 impl Docker {
+    /// Constructs a new Docker instance based on uri.
+    ///
+    /// uri can be either a unix or tcp socket, for example `unix:///run/docker.sock` or for tcp
+    /// `tcp://127.0.0.1:80`.
+    ///
+    /// Valid schemes are: `tcp`, `http`, `https`, `unix`
+    pub fn new<S>(uri: S) -> Result<Docker>
+    where
+        S: Into<String>,
+    {
+        let uri_str = uri.into();
+
+        #[allow(clippy::single_match)] // this match is much nicer than two separate if lets
+        match uri_str.strip_prefix("unix://") {
+            #[cfg(feature = "unix-socket")]
+            Some(path) => return Ok(Docker::unix(path)),
+            #[cfg(not(feature = "unix-socket"))]
+            Some(_) => return Err(Error::UnsupportedScheme { scheme: "unix" }),
+            _ => {}
+        }
+
+        let uri = uri_str.parse::<Uri>()?;
+
+        if let Some(scheme) = uri.scheme() {
+            match scheme.as_str() {
+                "tcp" | "http" | "https" => Docker::tcp(uri_str),
+                s => Err(Error::UnsupportedScheme { scheme: s.into() }),
+            }
+        } else {
+            Err(Error::UnsupportedScheme { scheme: "".into() })
+        }
+    }
+
     /// constructs a new Docker instance for a docker host listening at a url specified by an env var `DOCKER_HOST`,
-    /// falling back on unix:///var/run/docker.sock
-    pub fn new() -> Docker {
+    /// falling back to [DEFAULT_SOCKET_PATH](DEFAULT_SOCKET_PATH))
+    pub fn from_env() -> Result<Docker> {
         match env::var("DOCKER_HOST").ok() {
-            Some(host) => {
-                #[cfg(feature = "unix-socket")]
-                if let Some(path) = host.strip_prefix("unix://") {
-                    return Docker::unix(path);
-                }
-                let host = host.parse().expect("invalid url");
-                Docker::host(host)
-            }
+            Some(host) => Docker::new(host),
             #[cfg(feature = "unix-socket")]
-            None => Docker::unix("/var/run/docker.sock"),
+            None => Ok(Docker::unix(DEFAULT_SOCKET_PATH)),
             #[cfg(not(feature = "unix-socket"))]
-            None => panic!("Unix socket support is disabled"),
+            None => Err(Error::UnsupportedSocket { socket: "unix" }),
         }
     }
 
     /// Creates a new docker instance for a docker host
     /// listening on a given Unix socket.
     #[cfg(feature = "unix-socket")]
-    pub fn unix<S>(socket_path: S) -> Docker
+    fn unix<S>(socket_path: S) -> Docker
```

Comment:
Why was this made private?

Reply:
I'd also like to keep these public for cases users would like a more explicit constructor

Reply:
Wasn't sure about this, but I agree keeping them public is a better choice.